src/components/30-organisms/carousel/index.js

```diff
@@ -0,0 +1,228 @@
+    }, 50);
+  }
+
+  _nextSlide() {
+    let nextSlideIndex = this.visibleSlideIndex + 1;
+
+    if (nextSlideIndex > this.slides.length - 1) {
+      nextSlideIndex = 0;
+    }
+
+    this._setSlideVisibleWithAnimation(nextSlideIndex, ANIMATION_RIGHT_CLASS);
+  }
+
+  _previousSlide() {
+    let nextSlideIndex = this.visibleSlideIndex - 1;
+
+    if (nextSlideIndex < 0) {
+      nextSlideIndex = this.slides.length - 1;
+    }
+
+    this._setSlideVisibleWithAnimation(nextSlideIndex, ANIMATION_LEFT_CLASS);
+  }
+
+  _calculateContainerMinHeight() {
+    // we need to set carousel min height in case there are elements with different heights.
+    this.slides.forEach(node => {
+      node.style.display = 'block'; // not all the elements have initially a height > 0
+      if (node.clientHeight > this._carouselMinHeight) {
+        this._carouselMinHeight = node.clientHeight;
+      }
+    });
+  }
+
+  _onResize = debounce(() => {
+    this._carouselMinHeight = 0;
+    this._calculateContainerMinHeight();
+    this._setSlideVisibleWithAnimation(this.visibleSlideIndex, '');
+  }, 200);
+
+  // AutoRotate:
+
+  _startAutoRotate() {
+    if (!this.autorotatedisabled) {
+      this.autoRotateTimerID = setInterval(() => {
+        this._nextSlide();
+      }, this.autorotatetime);
+    }
+  }
+
+  _stopAutoRotate() {
+    clearInterval(this.autoRotateTimerID);
+  }
+
+  // Swipe for mobile devices:
+
+  _onSwipeLeft = () => {
+    this.handleNextButtonClick();
+  };
+
+  _onSwipeRight = () => {
+    this.handlePreviousButtonClick();
+  };
+
+  _initSwipe(usedElement = this) {
+    // usedElement as a property to encapsulate for unit tests
+    this.swiper = new Swipe(usedElement, usedElement._onSwipeLeft, usedElement._onSwipeRight);
+    this.swiper.run();
+  }
+
+  _terminateSwipe() {
+    if (this.swiper && this.swiper.stop) {
+      this.swiper.stop();
+    }
+  }
+
+  // Key Navigation:
+
+  _initKeyNavigation(usedDocument = this.ownerDocument) {
+    // usedDocument as a property to encapsulate for unit tests
+    if (this.keysenabled) {
+      usedDocument.addEventListener('keyup', this._handleKeyUp);
+    }
+  }
+
+  _terminateKeyNavigation(usedDocument = this.ownerDocument) {
+    // usedDocument as a property to encapsulate for unit tests
+    usedDocument.removeEventListener('keyup', this._handleKeyUp);
+  }
+
+  _handleKeyUp = ev => {
+    const e = ev || window.event;
+
+    if (e.keyCode === 37) {
+      this.handlePreviousButtonClick();
+    } else if (e.keyCode === 39) {
+      this.handleNextButtonClick();
+    }
+  };
+
+  constructor() {
+    super();
+    // Props
+    this.autorotatedisabled = false;
+    this.autorotatetime = 5000;
+    this.keysenabled = false;
+    this._animationWrapperClass = '';
+    this._carouselMinHeight = 0;
+    // Internal
+    this.autoRotateTimerID = null;
+    this.slides = null;
+    this.visibleSlideIndex = 0;
+    this.swiper = null;
+  }
+
+  firstUpdated() {
```

Comment:
cafefull, this function is called ONLY at the first render of the Component. If props changes then i have my doubt that it keeps working. I would add this event bounding in the render method to be secure that on each render correct event listener are set to dom node

Reply:
Im not entirely sure, maybe test it before refactoring

Reply:
everything works as expected if props changed. other components do also have event binding here in firstUpdated.